app/src/ui/preferences/advanced.tsx

```diff
@@ -48,38 +43,31 @@ export class Advanced extends React.Component<
     }
   }
 
-  public async componentDidMount() {
-    const [availableEditors, availableShells] = await Promise.all([
-      getAvailableEditors(),
-      getAvailableShells(),
-    ])
-
-    const editors = availableEditors.map(editor => editor.editor)
-    let selectedExternalEditor = this.props.selectedExternalEditor
+  public async componentWillReceiveProps(nextProps: IAdvancedPreferencesProps) {
```

Comment:
A quick note about what I stumbled on as part of #2551 and the preferences dialog:

 - `Preferences.tsx` (parent) has a `componentWillMount` function, and `Advanced.tsx` (child) had a `componentDidMount` function
 - In theory, I was hoping to see these functions fire in order relative to the hierarchy, but here is what I was seeing at runtime:
   - `Advanced.componentDidMount: selecting new editor: 'Visual Studio Code'`
   - `Preferences.onSelectedEditorChanged: got 'Visual Studio Code'`
   - `Preferences.componentWillMount: setting state: 'Atom'` 
 - Looks like any state change from child is discarded due to it being too early in the lifecycle 

So rather than hacking around this constraint, I've moved the state around computing the editor/shell list up into `Preferences` and used `componentWillReceiveProps` (later in the component lifecycle) to flow changes back when there's a mismatch between the selected editor and the list of available editors.